README.md
[![Build Status](http://ci.digitalbazaar.com/buildStatus/icon?job=bedrock-key-http)](http://ci.digitalbazaar.com/job/bedrock-key-http)

# bedrock-key-http

## Requirements

- npm v3+


CHANGELOG.md
# bedrock-key-http ChangeLog

## 2.2.1 - 2018-03-26

### Changed
- Update `bedrock-mongodb` peer dependency.

## 2.2.0 - 2017-06-27

### Changed
- Upgrade `bedrock-key` peer dependency from 3.x to 4.x.

## 2.1.3 - 2017-04-18

### Changed
- Add test suite.

## 2.1.2 - 2016-08-28

### Changed
- Add doc configs from bedrock-docs.

## 2.1.1 - 2016-06-07

### Changed
- Update dependencies.

## 2.1.0 - 2016-05-27

### Added
- Add `capability` filter to keys query.

## 2.0.2 - 2016-04-28

### Changed
- Update bedrock-key dependency.

## 2.0.1 - 2016-04-15

### Changed
- Update bedrock dependencies.

## 2.0.0 - 2016-03-03

### Changed
- Update package dependencies for npm v3 compatibility.

## 1.0.1 - 2016-02-01

### Changed
- Add main keys collection.

## 1.0.0 - 2016-01-31

### Changed
- See git history for changes.


test/test.js
/*
 * Copyright (c) 2016-2017 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
require('bedrock-key-http');

require('bedrock-test');
bedrock.start();


test/test.config.js
/*
 * Bedrock Key HTTP Module Test Configuration.
 *
 * Copyright (c) 2012-2017 Digital Bazaar, Inc. All rights reserved.
 */
const config = require('bedrock').config;
const path = require('path');

const permissions = config.permission.permissions;
const roles = config.permission.roles;

config.mocha.tests.push(path.join(__dirname, 'mocha'));

// MongoDB
config.mongodb.name = 'bedrock_key_http_test';
config.mongodb.local.collection = 'bedrock_key_http_test';
config.mongodb.dropCollections = {};
config.mongodb.dropCollections.onInit = true;
config.mongodb.dropCollections.collections = [];

roles['bedrock-key-http.test'] = {
  id: 'bedrock-key-http.test',
  label: 'Key HTTP Test Role',
  comment: 'Role for Test User',
  sysPermission: [
    permissions.PUBLIC_KEY_REMOVE.id,
    permissions.PUBLIC_KEY_ACCESS.id,
    permissions.PUBLIC_KEY_CREATE.id,
    permissions.PUBLIC_KEY_EDIT.id
  ]
};


test/mocha/helpers.js
/*
 * Copyright (c) 2016-2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const async = require('async');
const brIdentity = require('bedrock-identity');
const brKey = require('bedrock-key');
const database = require('bedrock-mongodb');
const uuid = require('uuid').v4;

const api = {};
module.exports = api;

api.createHttpSignatureRequest = function(options) {
  const newRequest = {
    url: options.url,
    httpSignature: {
      key: options.identity.keys.privateKey.privateKeyPem,
      keyId: options.identity.keys.publicKey.id,
      headers: ['date', 'host', 'request-line']
    }
  };
  if(options.body) {
    newRequest.body = options.body;
  }
  return newRequest;
};

api.createIdentity = function(userName) {
  const newIdentity = {
    id: 'did:' + uuid(),
    type: 'Identity',
    sysSlug: userName,
    label: userName,
    email: userName + '@bedrock.dev',
    sysPassword: 'password',
    sysPublic: ['label', 'url', 'description'],
    sysResourceRole: [],
    url: 'https://example.com',
    description: userName,
    sysStatus: 'active'
  };
  return newIdentity;
};

api.removeCollection = function(collection, callback) {
  const collectionNames = [collection];
  database.openCollections(collectionNames, () => {
    async.each(collectionNames, function(collectionName, callback) {
      database.collections[collectionName].remove({}, callback);
    }, function(err) {
      callback(err);
    });
  });
};

api.removeCollections = function(callback) {
  const collectionNames = ['identity', 'eventLog', 'publicKey'];
  database.openCollections(collectionNames, () => {
    async.each(collectionNames, (collectionName, callback) => {
      database.collections[collectionName].remove({}, callback);
    }, function(err) {
      callback(err);
    });
  });
};

api.createKeyPair = function(options) {
  const userName = options.userName;
  const publicKey = options.publicKey;
  const privateKey = options.privateKey;
  let ownerId = null;
  if(userName === 'userUnknown') {
    ownerId = '';
  } else {
    ownerId = options.userId;
  }
  const newKeyPair = {
    publicKey: {
      '@context': 'https://w3id.org/identity/v1',
      id: ownerId + '/keys/1',
      type: 'CryptographicKey',
      owner: ownerId,
      label: 'Signing Key 1',
      publicKeyPem: publicKey
    },
    privateKey: {
      type: 'CryptographicKey',
      owner: ownerId,
      label: 'Signing Key 1',
      publicKey: ownerId + '/keys/1',
      privateKeyPem: privateKey
    }
  };
  return newKeyPair;
};

api.prepareDatabase = function(mockData, callback) {
  async.series([
    callback => {
      api.removeCollections(callback);
    },
    callback => {
      insertTestData(mockData, callback);
    }
  ], callback);
};

// Insert identities and public keys used for testing into database
function insertTestData(mockData, callback) {
  async.forEachOf(mockData.identities, function(identity, key, callback) {
    async.parallel([
      function(callback) {
        brIdentity.insert(null, identity.identity, callback);
      },
      function(callback) {
        brKey.addPublicKey(null, identity.keys.publicKey, callback);
      }
    ], callback);
  }, function(err) {
    if(err) {
      if(!database.isDuplicateError(err)) {
        // duplicate error means test data is already loaded
        return callback(err);
      }
    }
    callback();
  }, callback);
}


lib/main.js
/*
 * Bedrock Key HTTP Module
 *
 * Copyright (c) 2012-2016 Digital Bazaar, Inc. All rights reserved.
 */
var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var brKey = require('bedrock-key');
var brPassport = require('bedrock-passport');
var brRest = require('bedrock-rest');
var database = require('bedrock-mongodb');
var cors = require('cors');
var docs = require('bedrock-docs');
var util = require('util');

var BedrockError = bedrock.util.BedrockError;
var ensureAuthenticated = brPassport.ensureAuthenticated;
var validate = require('bedrock-validation').validate;

var logger = bedrock.loggers.get('app');

require('./config');

// setup path params
bedrock.events.on('bedrock-express.init', function(app) {
  app.param('publicKey', brRest.idParam);
});

// add routes
bedrock.events.on('bedrock-express.configure.routes', addRoutes);

function addRoutes(app) {
  var basePath = bedrock.config.key.basePath;
  var identityBasePath;
  // FIXME: deprecated old key path
  if('idp' in bedrock.config) {
    identityBasePath = bedrock.config.idp.identityBasePath + '/:identity';
  }

  // TODO: determine if this API can be used to create DID-based keys
  app.post(basePath,
    ensureAuthenticated,
    validate('services.key.postKeys'),
    function(req, res, next) {
      var owner;
      if('owner' in req.body) {
        logger.warning('FIXME: add checks when adding public key with owner');
        owner = req.body.owner;
      } else {
        owner = req.user.identity.id;
      }

      // build public key
      var publicKey = {
        '@context': bedrock.config.constants.IDENTITY_CONTEXT_V1_URL,
        type: 'CryptographicKey',
        owner: owner,
        label: req.body.label,
        publicKeyPem: req.body.publicKeyPem
      };

      var privateKey = null;
      if('privateKeyPem' in req.body) {
        privateKey = {
          type: 'CryptographicKey',
          owner: owner,
          label: req.body.label,
          privateKeyPem: req.body.privateKeyPem
        };
      }

      // add public key
      brKey.addPublicKey(
        req.user.identity, publicKey, privateKey, function(err) {
        if(err && database.isDuplicateError(err)) {
          return next(new BedrockError(
            'The identity key is a duplicate and could not be added.',
            'DuplicateIdentityKey', {
              httpStatusCode: 409,
              'public': true
            }));
        }
        if(err) {
          return next(new BedrockError(
            'The identity key could not be added.',
            'AddIdentityKeyFailed', {
              httpStatusCode: 400,
              'public': true
            }, err));
        }
        // return key
        res.set('Location', publicKey.id);
        res.status(201).json(publicKey);
      });
    });
  docs.annotate.post(basePath, {
    description: 'Associate a public key with the identity.',
    securedBy: ['cookie', 'hs1'],
    schema: 'services.key.postKeys',
    responses: {
      201: 'Key registration was successful.',
      400: 'The key could not be added.',
      409: 'The key is a duplicate and was not added.'
    }
  });

  // FIXME: 404 if owner unknown vs []?
  // TODO: determine if DID-based keys should be returned
  // API for getting all keys (including DID-based keys?)
  app.options(basePath, cors());
  app.get(basePath,
    cors(),
    validate({
      query: 'services.key.getKeysQuery'
    }),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var actor = req.user ? req.user.identity : undefined;
        brKey.getPublicKeys(
          req.query.owner, actor, req.query, function(err, records) {
          if(err) {
            return callback(err);
          }
          callback(err, records ? _.map(records, 'publicKey') : null);
        });
      }
    }));
  // FIXME add owner param docs
  docs.annotate.get(basePath, {
    description: 'Get the list of public keys associated with an identity.',
    securedBy: ['null', 'cookie', 'hs1'],
    responses: {
      200: {
        'application/ld+json': {
          'example': 'examples/get.identity.keys.jsonld'
        }
      }
    }
  });

  function _postPublicKey(publicKeyId, req, res, next) {
    if(publicKeyId !== req.body.id) {
      // id mismatch
      return next(new BedrockError(
        'Incorrect key id.',
        'KeyIdError', {
          'public': true,
          httpStatusCode: 400
        }));
    }

    if('revoked' in req.body) {
      // revoke public key
      return brKey.revokePublicKey(
        req.user.identity, publicKeyId, function(err, key) {
          if(err) {
            return next(err);
          }
          res.status(200).send(key);
        });
    }

    async.waterfall([
      function(callback) {
        brKey.getPublicKey({id: publicKeyId}, function(err, publicKey) {
          callback(err, publicKey);
        });
      },
      function(key, callback) {
        // update public key
        if('label' in req.body) {
          key.label = req.body.label;
        }
        brKey.updatePublicKey(req.user.identity, key, callback);
      }
    ], function(err) {
      if(err) {
        return next(err);
      }
      res.sendStatus(204);
    });
  }

  // FIXME: support for deprecated old key path
  function _createOldPublicKeyId(req) {
    return util.format('%s%s/%s/keys/%s',
      bedrock.config.server.baseUri,
      bedrock.config.idp.identityBasePath,
      encodeURIComponent(req.params.identity),
      encodeURIComponent(req.params.publicKey));
  }

  // API for updating a local public key
  app.post(basePath + '/:publicKey',
    ensureAuthenticated,
    validate('services.key.postKey'),
    function(req, res, next) {
      // get ID from URL
      var publicKeyId = brKey.createPublicKeyId(req.params.publicKey);
      _postPublicKey(publicKeyId, req, res, next);
    });
  // FIXME: deprecated old key path support
  if(identityBasePath) {
    app.post(identityBasePath + '/keys/:publicKey',
      ensureAuthenticated,
      validate('services.key.postKey'),
      function(req, res, next) {
        // get old ID from URL
        var publicKeyId = _createOldPublicKeyId(req);
        _postPublicKey(publicKeyId, req, res, next);
      });
  }
  docs.annotate.post(basePath + '/:publicKey', {
    description: 'Modify an existing public key.',
    securedBy: ['cookie', 'hs1'],
    schema: 'services.key.postKey',
    responses: {
      200: 'The key was revoked successfully.',
      204: 'The key was updated successfully.',
      400: 'The key could not be modified.'
    }
  });

  // API for getting a local public key
  app.options(basePath + '/:publicKey', cors());
  app.get(basePath + '/:publicKey',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var publicKeyId = brKey.createPublicKeyId(req.params.publicKey);
        // get public key
        brKey.getPublicKey({id: publicKeyId}, function(err, key) {
          callback(err, key);
        });
      }
    }));
  // FIXME: deprecated old key path support
  if(identityBasePath) {
    app.options(identityBasePath + '/keys/:publicKey', cors());
    app.get(identityBasePath + '/keys/:publicKey',
      cors(),
      brRest.when.prefers.ld,
      brRest.linkedDataHandler({
        get: function(req, res, callback) {
          // get old ID from URL
          var publicKeyId = _createOldPublicKeyId(req);
          // get public key
          brKey.getPublicKey({id: publicKeyId}, function(err, key) {
            callback(err, key);
          });
        }
      }));
  }
  docs.annotate.get(basePath + '/:publicKey', {
    description: 'Get a public keys associated with an identity.',
    securedBy: ['null', 'cookie', 'hs1'],
    responses: {
      200: {
        'application/ld+json': {
          'example': 'examples/get.identity.keys.publicKey.jsonld'
        }
      },
      404: 'The key was not found.'
    }
  });
}

bedrock.events.on('bedrock-views.vars.get', addViewVars);

function addViewVars(req, vars, callback) {
  // FIXME which namespaces?
  vars['key-http'] = {};
  vars['key-http'].basePath = bedrock.config.key.basePath;
  vars['key-http'].baseUri = vars.baseUri + vars['key-http'].basePath;

  callback();
}


schemas/services.key.js
/*
 * Copyright (c) 2012-2015 Digital Bazaar, Inc. All rights reserved.
 */
var constants = require('bedrock').config.constants;
var schemas = require('bedrock-validation').schemas;

var postKey = {
  type: 'object',
  properties: {
    '@context': schemas.jsonldContext(constants.IDENTITY_CONTEXT_V1_URL),
    id: schemas.identifier(),
    label: schemas.label({required: false}),
    revoked: {
      required: false,
      type: 'string'
    }
  },
  additionalProperties: false
};

var getKeysQuery = {
  title: 'Get Keys Query',
  type: 'object',
  properties: {
    owner: schemas.identifier({required: true}),
    capability: {
      required: false,
      enum: ['sign']
    }
  },
  additionalProperties: false
};

var postKeys = {
  type: 'object',
  properties: {
    '@context': schemas.jsonldContext(constants.IDENTITY_CONTEXT_V1_URL),
    label: schemas.label(),
    owner: schemas.identifier({
      required: false
    }),
    publicKeyPem: schemas.publicKeyPem(),
    privateKeyPem: schemas.privateKeyPem({required: false})
  },
  additionalProperties: false
};

module.exports.postKey = function() {
  return postKey;
};
module.exports.getKeysQuery = function() {
  return getKeysQuery;
};
module.exports.postKeys = function() {
  return postKeys;
};


lib/index.js
/*
 * Bedrock Key HTTP Module
 *
 * Copyright (c) 2012-2016 Digital Bazaar, Inc. All rights reserved.
 */
var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var brKey = require('bedrock-key');
var brPassport = require('bedrock-passport');
var brRest = require('bedrock-rest');
var database = require('bedrock-mongodb');
var cors = require('cors');
var docs = require('bedrock-docs');
var util = require('util');

var BedrockError = bedrock.util.BedrockError;
var ensureAuthenticated = brPassport.ensureAuthenticated;
var validate = require('bedrock-validation').validate;

var logger = bedrock.loggers.get('app');

require('./config');

// setup path params
bedrock.events.on('bedrock-express.init', function(app) {
  app.param('publicKey', brRest.idParam);
});

// add routes
bedrock.events.on('bedrock-express.configure.routes', addRoutes);

function addRoutes(app) {
  var keyPath = bedrock.config['key-http'].routes.basePath;
  var identityBasePath;
  // FIXME: deprecated old key path
  if('idp' in bedrock.config) {
    identityBasePath = bedrock.config.idp.identityBasePath + '/:identity';
  }

  // TODO: determine if this API can be used to create DID-based keys
  app.post(keyPath,
    ensureAuthenticated,
    validate('services.key.postKeys'),
    function(req, res, next) {
      var owner;
      if('owner' in req.body) {
        logger.warning('FIXME: add checks when adding public key with owner');
        owner = req.body.owner;
      } else {
        owner = req.user.identity.id;
      }

      // build public key
      var publicKey = {
        '@context': bedrock.config.constants.IDENTITY_CONTEXT_V1_URL,
        type: 'CryptographicKey',
        owner: owner,
        label: req.body.label,
        publicKeyPem: req.body.publicKeyPem
      };

      var privateKey = null;
      if('privateKeyPem' in req.body) {
        privateKey = {
          type: 'CryptographicKey',
          owner: owner,
          label: req.body.label,
          privateKeyPem: req.body.privateKeyPem
        };
      }

      // add public key
      brKey.addPublicKey(
        req.user.identity, publicKey, privateKey, function(err) {
          if(err && database.isDuplicateError(err)) {
            return next(new BedrockError(
              'The identity key is a duplicate and could not be added.',
              'DuplicateIdentityKey', {
                httpStatusCode: 409,
                'public': true
              }));
          }
          if(err) {
            return next(new BedrockError(
              'The identity key could not be added.',
              'AddIdentityKeyFailed', {
                httpStatusCode: 400,
                'public': true
              }, err));
          }
          // return key
          res.set('Location', publicKey.id);
          res.status(201).json(publicKey);
        });
    });
  docs.annotate.post(keyPath, {
    description: 'Associate a public key with the identity.',
    securedBy: ['cookie', 'hs1'],
    schema: 'services.key.postKeys',
    responses: {
      201: 'Key registration was successful.',
      400: 'The key could not be added.',
      409: 'The key is a duplicate and was not added.'
    }
  });

  // FIXME: 404 if owner unknown vs []?
  // TODO: determine if DID-based keys should be returned
  // API for getting all keys (including DID-based keys?)
  app.options(keyPath, cors());
  app.get(keyPath,
    cors(),
    validate({
      query: 'services.key.getKeysQuery'
    }),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var actor = req.user ? req.user.identity : undefined;
        brKey.getPublicKeys(
          req.query.owner, actor, req.query, function(err, records) {
            if(err) {
              return callback(err);
            }
            callback(err, records ? _.map(records, 'publicKey') : null);
          });
      }
    }));
  // FIXME add owner param docs
  docs.annotate.get(keyPath, {
    description: 'Get the list of public keys associated with an identity.',
    securedBy: ['null', 'cookie', 'hs1'],
    responses: {
      200: {
        'application/ld+json': {
          'example': 'examples/get.identity.keys.jsonld'
        }
      }
    }
  });

  function _postPublicKey(publicKeyId, req, res, next) {
    if(publicKeyId !== req.body.id) {
      // id mismatch
      return next(new BedrockError(
        'Incorrect key id.',
        'KeyIdError', {
          'public': true,
          httpStatusCode: 400
        }));
    }

    if('revoked' in req.body) {
      // revoke public key
      return brKey.revokePublicKey(
        req.user.identity, publicKeyId, function(err, key) {
          if(err) {
            return next(err);
          }
          res.status(200).send(key);
        });
    }

    async.waterfall([
      function(callback) {
        brKey.getPublicKey({id: publicKeyId}, function(err, publicKey) {
          callback(err, publicKey);
        });
      },
      function(key, callback) {
        // update public key
        if('label' in req.body) {
          key.label = req.body.label;
        }
        brKey.updatePublicKey(req.user.identity, key, callback);
      }
    ], function(err) {
      if(err) {
        return next(err);
      }
      res.sendStatus(204);
    });
  }

  // FIXME: support for deprecated old key path
  function _createOldPublicKeyId(req) {
    return util.format('%s%s/%s/keys/%s',
      bedrock.config.server.baseUri,
      bedrock.config.idp.identityBasePath,
      encodeURIComponent(req.params.identity),
      encodeURIComponent(req.params.publicKey));
  }

  // API for updating a local public key
  app.post(keyPath + '/:publicKey',
    ensureAuthenticated,
    validate('services.key.postKey'),
    function(req, res, next) {
      // get ID from URL
      var publicKeyId = brKey.createPublicKeyId(req.params.publicKey);
      _postPublicKey(publicKeyId, req, res, next);
    });
  // FIXME: deprecated old key path support
  if(identityBasePath) {
    app.post(identityBasePath + '/keys/:publicKey',
      ensureAuthenticated,
      validate('services.key.postKey'),
      function(req, res, next) {
        // get old ID from URL
        var publicKeyId = _createOldPublicKeyId(req);
        _postPublicKey(publicKeyId, req, res, next);
      });
  }
  docs.annotate.post(keyPath + '/:publicKey', {
    description: 'Modify an existing public key.',
    securedBy: ['cookie', 'hs1'],
    schema: 'services.key.postKey',
    responses: {
      200: 'The key was revoked successfully.',
      204: 'The key was updated successfully.',
      400: 'The key could not be modified.'
    }
  });

  // API for getting a local public key
  app.options(keyPath + '/:publicKey', cors());
  app.get(keyPath + '/:publicKey',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var publicKeyId = brKey.createPublicKeyId(req.params.publicKey);
        // get public key
        brKey.getPublicKey({id: publicKeyId}, function(err, key) {
          callback(err, key);
        });
      }
    }));
  // FIXME: deprecated old key path support
  if(identityBasePath) {
    app.options(identityBasePath + '/keys/:publicKey', cors());
    app.get(identityBasePath + '/keys/:publicKey',
      cors(),
      brRest.when.prefers.ld,
      brRest.linkedDataHandler({
        get: function(req, res, callback) {
          // get old ID from URL
          var publicKeyId = _createOldPublicKeyId(req);
          // get public key
          brKey.getPublicKey({id: publicKeyId}, function(err, key) {
            callback(err, key);
          });
        }
      }));
  }
  docs.annotate.get(keyPath + '/:publicKey', {
    description: 'Get a public keys associated with an identity.',
    securedBy: ['null', 'cookie', 'hs1'],
    responses: {
      200: {
        'application/ld+json': {
          'example': 'examples/get.identity.keys.publicKey.jsonld'
        }
      },
      404: 'The key was not found.'
    }
  });
}

bedrock.events.on('bedrock-views.vars.get', addViewVars);

function addViewVars(req, vars, callback) {
  // FIXME which namespaces?
  vars['key-http'] = {};
  vars['key-http'].basePath = bedrock.config['key-http'].routes.basePath;
  vars['key-http'].baseUri = vars.baseUri + vars['key-http'].basePath;

  callback();
}
